### PR TITLE
ENH: add test for x in op.domain in all solvers, closes #291

### LINKED ISSUE #291: X in space checks for solvers
The solvers currently have very lightweight checks, it would occationally be nice to have these improved so we properly check input, that way we dont get cryptic messages and need to spend time figuring out what went wrong.

### PR DESCRIPTION
**Pull request by adler-j**
Added input sanitation to the solvers.

### CODE REVIEW

**Comment by kohr-h** [comment]
Good initiative. A few comments:

- When we raise `TypeError`, we should print the `repr` of at least the object that caused the error, so the user sees the provided type against the expected one in the error message. We could also just print the type, which would be a valid alternative.
- The Newton-type methods should not expect something named `grad` by default, that's a specific case. Now there was this inconsistency that he quasi-Newton ones take `grad` and Newton takes `op`, but IMO this inconsistency should be resolved in the other direction. Actually, I would prefer to keep the names untouched in this PR and do the restructuring separately, see #52.

**Comment by adler-j** [comment]
Fixed according to comments. Merge?

File: odl/solvers/advanced/chambolle_pock.py [hunk 1]
```diff
@@ -148,13 +148,13 @@ def chambolle_pock_solver(op, x, tau, sigma, proximal_primal, proximal_dual,
     """
     # Forward operator
     if not isinstance(op, Operator):
-        raise TypeError('`op` {} is not an instance of {}'
-                        ''.format(op, Operator))
+        raise TypeError('`op` {!r} is not an `Operator`'
```
Comment (kohr-h) [inline odl/solvers/advanced/chambolle_pock.py]: We usually write `<type> instance`. Debatable for sure, but consistency is almost always good :-)

**Comment by kohr-h** [comment]
See above, otherwise mergable.

**Pull request merged**